### ISSUE TITLE
Provide an ability for analyzer plugins to contribute to `dart fix` fixes set

### ISSUE THREAD

**Issue by incendial** [dart-lang/sdk#46760]
Following up https://github.com/Dart-Code/Dart-Code/issues/3502#issuecomment-889365207.

Dart Code Metrics plugin provides some lint rules which can be easily auto-fixed and this feature would be very useful for our users, especially after [the fix-all](https://dartcode.org/releases/v3-25/#fix-all-command) IDE command was introduced. We even got a related question quite some time ago: https://github.com/dart-code-checker/dart-code-metrics/issues/195.

Let me know if I can help somehow.

**Comment by bwilkerson** [comment]
@devoncarew @csells for visibility

**Comment by csells** [comment]
What's involved in letting others contribute lint's?

**Comment by bwilkerson** [comment]
Plugins built using the `analyzer_plugin` package can already contribute lints. They can also contribute fixes associated with a diagnostic, and those fixes can either fix a single occurrence of the diagnostic or can fix all occurrences in the file.

In order to allow plugins to contribute fixes that can be applied using `dart fix` we would minimally need to
- enhance the protocol used by server to communicate with plugins (to allow server to request such fixes from plugins),
- enhance the code that implements `dart fix` to run and communicate with plugins, and
- figure out how to merge the changes proposed by server and those proposed by plugins in such a way as to prevent those change from corrupting the user's code because of conflicts, and then implement the merge operation.

We also have some support classes in the analysis server that would probably be useful to move into the `analyzer_plugin` package so that it's easier for plugin authors to create the fixes.

It doesn't sound like a prohibitively large amount of work, but definitely something we'll need to schedule.

**Comment by csells** [comment]
As a developer consuming lints from a 3rd party, ie not from the Dart or Flutter SDKs, do I have the ability to turn the lints off in my settings so they don't get used in 'fix all'? Or, to ask another way, in a turn a lint off in my analysis options, does that exclude it from 'fix all'?

**Comment by bwilkerson** [comment]
The intended behavior of `dart fix` is that it will only suggest a change if it's a fix for a diagnostic. That's the way the analysis server currently works, and if plugins follow that expectation then yes, disabling a lint will prevent any fixes for violations of the lint from being applied. There's nothing to stop a malicious plugin from suggesting changes that are not associated with a fix, but that's because we assume that plugin authors aren't malicious.

**Comment by csells** [comment]
Mmm... Malicious plug-in authors automatically changing my code isn't something I'd like to enable.

**Comment by bwilkerson** [comment]
Given that users have to explicitly enable the use of a plugin (and can disable it at any time) I don't think it's very likely to be a problem. Besides, source control should be able to guard against any permanent damage.

**Comment by csells** [comment]
fair

**Comment by incendial** [comment]
> figure out how to merge the changes proposed by server and those proposed by plugins in such a way as to prevent those change from corrupting the user's code because of conflicts, and then implement the merge operation.

I'm not sure that it's a good suggestion, but from a plugin author standpoint it'd be easier to manage plugin changes separately from the server ones. With this approach changes won't conflict with each other and I as a plugin author will be responsible only for my plugin fixes not to conflict with each other.

**Comment by bwilkerson** [comment]
> ... from a plugin author standpoint it'd be easier to manage plugin changes separately from the server ones.

Absolutely. I'd like to find a way for plugin authors to not need to know about the changes from the server nor from any other plugins.

> With this approach changes won't conflict with each other ...

I'm not sure what approach you're referring to, but it isn't clear to me that there's any approach that can guarantee that there won't be collisions. For example, if the user has enabled a lint of the form 'use the method M2 rather than the method M1' and the fix requires replacing the argument list as well as the name of the method, and a plugin has a check of the form 'the first argument of M1 should always be wrapped in a call to M3' and the fix requires replacing the first argument, then the ranges of the edits will overlap, which is what causes the corruption.

For server we're making use of some newly added functionality in the change builders. If an edit is created that conflicts with a previously created edit then an exception is thrown. In the outer loop, we create a fallback position prior to attempting each fix, then if the exception is thrown by the fix code we drop back to the earlier known good state and skip the fix that caused the conflict.

This isn't ideal in the sense that some fixes won't be applied, but it does prevent corruption of the user's code.

My first guess at a process is that plugins would want to send back one set of edits for each diagnostic being addressed (so each `SourceChange` is a collection of edits that need to be applied atomically) and that server would then attempt to merge those edits, checking for conflicts with previous edits and discarding those edits that conflict, much the way it handles its own fixes.

In addition to the existing issues with our current approach, this adds the problem that the order in which the edits are applied is potentially visible to the user because fixes from plugins will be increasingly likely to be discarded due to conflicts. One mitigation for this is to improve server's ability to merge edits; it does a little work now, but there's plenty of room for improvement. (It does, however, have the nice property that plugins don't need to worry about whether the fixes they create are conflicting because that would be handled for them by server.)

**Comment by incendial** [comment]
> I'm not sure what approach you're referring to, but it isn't clear to me that there's any approach that can guarantee that there won't be collisions. For example, if the user has enabled a lint of the form 'use the method M2 rather than the method M1' and the fix requires replacing the argument list as well as the name of the method, and a plugin has a check of the form 'the first argument of M1 should always be wrapped in a call to M3' and the fix requires replacing the first argument, then the ranges of the edits will overlap, which is what causes the corruption.

Sorry for not being clear. Initially I thought about applying fixes sequentially, for example, server fixes at first and then each plugin fixes separately. That will not create a situation when a plugin fixes are getting reverted because of a conflict with server fixes, but only if they somehow conflict with each other. But the more I think about that the more I realize how slow this can be since we'll need to apply server fixes and then recalculate possible fixes for the next plugin and apply them. And it will go until all plugins provided theirs fixes.

> In addition to the existing issues with our current approach, this adds the problem that the order in which the edits are applied is potentially visible to the user because fixes from plugins will be increasingly likely to be discarded due to conflicts. One mitigation for this is to improve server's ability to merge edits; it does a little work now, but there's plenty of room for improvement.

That's why I thought about sequential application, but if there is plenty of room for improvement then I hope that merging can have as little amount of discards as sequential application.

**Comment by srawlins** [comment]
We've deprecated the legacy plugin system, as of Dart 3.14.2, and will be removing support for the legacy plugin system in an upcoming release of the Dart SDK. The [new plugins system](https://dart.dev/tools/analyzer-plugins) is supported and continually improving.

This specific feature is not yet supported in the new plugin system, but there is an open request for it.
